lib/integrations/azureoidc/accessgraph_sync.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization"
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/cloud/provisioning"
+	"github.com/gravitational/teleport/lib/config"
+	"github.com/gravitational/teleport/lib/msgraph"
+	tslices "github.com/gravitational/teleport/lib/utils/slices"
+)
+
+// graphAppId is the pre-defined application ID of the Graph API
+// Ref: [https://learn.microsoft.com/en-us/troubleshoot/entra/entra-id/governance/verify-first-party-apps-sign-in#application-ids-of-commonly-used-microsoft-applications].
+const graphAppId = "00000003-0000-0000-c000-000000000000"
+
+var requiredGraphRoleNames = []string{
+	"User.ReadBasic.All",
+	"Group.Read.All",
+	"Directory.Read.All",
+	"User.Read.All",
+	"Policy.Read.All",
+}
+
+func newManagedIdAction(cred *azidentity.DefaultAzureCredential, subId string, managedId string, roleName string) (*provisioning.Action, error) {
+	runnerFn := func(ctx context.Context) error {
+		// Create the role
+		roleDefCli, err := armauthorization.NewRoleDefinitionsClient(cred, nil)
+		if err != nil {
+			return trace.Wrap(fmt.Errorf("failed to create role definitions client: %w", err))
+		}
+		roleDefId := uuid.New().String()
+		customRole := "CustomRole"
+		scope := fmt.Sprintf("/subscriptions/%s", subId)
+		roleDefinition := armauthorization.RoleDefinition{
+			Name: &roleDefId,
+			Properties: &armauthorization.RoleDefinitionProperties{
+				RoleName: &roleName,
+				RoleType: &customRole,
+				Permissions: []*armauthorization.Permission{
+					{
+						Actions: tslices.ToPointers([]string{
+							"Microsoft.Compute/virtualMachines/read",
+							"Microsoft.Compute/virtualMachineScaleSets/virtualMachines/read",
+							"Microsoft.Authorization/roleDefinitions/read",
+							"Microsoft.Authorization/roleAssignments/read",
+						}),
+					},
+				},
+				AssignableScopes: []*string{&scope}, // Scope must be provided
+			},
+		}
+		roleRes, err := roleDefCli.CreateOrUpdate(ctx, scope, roleDefId, roleDefinition, nil)
+		if err != nil {
+			return trace.Wrap(fmt.Errorf("failed to create custom role: %w", err))
+		}
+
+		// Assign the new role to the managed identity
+		roleAssignCli, err := armauthorization.NewRoleAssignmentsClient(subId, cred, nil)
+		if err != nil {
+			return fmt.Errorf("failed to create role assignments client: %w", err)
+		}
+		assignName := uuid.New().String()
+		if err != nil {
+			return trace.Wrap(fmt.Errorf("failed to create role assignments client: %w", err))
+		}
+		roleAssignParams := armauthorization.RoleAssignmentCreateParameters{
+			Properties: &armauthorization.RoleAssignmentProperties{
+				PrincipalID:      &managedId,
+				RoleDefinitionID: roleRes.ID,
+			},
+		}
+		_, err = roleAssignCli.Create(ctx, scope, assignName, roleAssignParams, nil)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to assign role %s to principal %s: %w", roleName, managedId, err)
+		}
```

Comment:
Prefer shorter variable scopes when possible.
```suggestion
		if err := roleAssignCli.Create(ctx, scope, assignName, roleAssignParams, nil); err != nil {
			return fmt.Errorf(
				"failed to assign role %s to principal %s: %w", roleName, managedId, err)
		}
```